### PR TITLE
Post on sharing C++ code

### PR DESCRIPTION
**Pull request by pratikunterwegs**
This PR draft is for a proposed post on sharing C++ code between Rcpp packages.

### CODE REVIEW

**Comment by pratikunterwegs** [comment]
Hi @TimTaylor see work in progress on the post here. There's only a conclusion left to write, might be able to open this PR today, but if not feel free to take a look and add any suggestions.

File: posts/share-cpp/index.qmd [hunk 1]
```diff
@@ -14,12 +14,12 @@ format:
...
 
 Allowing C++ code to be shared can be very beneficial for the same reasons that sharing R code is --- packaging code is a reliable way to reuse it.
 
-Some widely used examples of this practice are the {RcppEigen}, {RcppArmadillo}, {RcppGSL}, and Boost Headers {BH} packages.
-Indeed, in the Epiverse-TRACE team, {RcppEigen} underpins the {finalsize} and upcoming {epidemics} packages.
+Some widely used examples of this practice are the [{RcppEigen}](https://cran.r-project.org/web/packages/RcppEigen/index.html), [{RcppArmadillo}](https://cran.r-project.org/web/packages/RcppArmadillo/index.html), [{RcppGSL}](https://cran.r-project.org/web/packages/RcppGSL/index.html), and [Boost Headers {BH}](https://cran.r-project.org/web/packages/BH/index.html) packages.
+Indeed, in the Epiverse-TRACE team, {RcppEigen} underpins the [{finalsize}](https://github.com/epiverse-trace/finalsize) and upcoming [{epidemics}](https://github.com/epiverse-trace/epidemics) packages.
 
 ## Two ways to share C++ code?
```
Comment (adamkucharski) [inline posts/share-cpp/index.qmd:25]: Worth numbering items in lines 28-30?

File: posts/share-cpp/index.qmd [hunk 2]
```diff
@@ -96,14 +96,14 @@ We would however recommend not autogenerating headers from Rcpp functions, but r
 There are few to no documented rules around the content of the package header.
```
Comment (adamkucharski) [inline posts/share-cpp/index.qmd:96]: In line 62 above, wasn't clear to me if that command generates a header file under `inst/include` AND a header file named `mypackage_RcppExports.h`, or if these are the same thing?

File: posts/share-cpp/index.qmd [hunk 3]
```diff
@@ -184,7 +184,7 @@ void print_hello_world() {
 
 ::: {.callout-tip}
 Remember to add `PKG_CPPFLAGS += -I../inst/include/` to both `Makevars` and `Makevars.win` under `src/`.
-Furthermore, as noted, the package will not automatically cause a rebuild when headers are modified --- this needs to be done manually.
+Furthermore, as noted in the [Rcpp attributes documentation](https://dirk.eddelbuettel.com/code/rcpp/Rcpp-attributes.pdf), the package will not automatically cause a rebuild when headers are modified --- this needs to be done manually.
 :::
 
 ## Linking header code between pacakges
```
Comment (adamkucharski) [inline posts/share-cpp/index.qmd:190]: Typo in line 197 (specifcy)?

File: posts/share-cpp/index.qmd [hunk 4]
```diff
@@ -230,11 +230,13 @@ cppcheck --std=c++14 --language=c++ --enable=warning,style --error-exitcode=1 in
...
-This is easy to do using the [Catch2 testing framework](link WIP), which is [conveniently available using {testthat}](link WIP) --- this is an extensive topic for another post.
-
-<!-- From the docs for Rcpp Attr:  -->
-<!-- https://gallery.rcpp.org/articles/mixing-modules-and-export/ -->
+This is easy to do using the [Catch2 testing framework](https://github.com/catchorg/Catch2), which is [conveniently available using {testthat}](https://testthat.r-lib.org/reference/use_catch.html) --- this is an extensive topic for another post.
 
 ## Conclusion WIP
 
-<!-- Three sentences here. -->
+Overall then, developing an Rcpp-based package with C++ code sharing in mind takes some organisation, or even reorganisation, of the C++ codebase.
```
Comment (adamkucharski) [inline posts/share-cpp/index.qmd:237]: I'd cut 'overall then', as conclusion heading already suggests this to reader (or shorten to just 'overall').

**Review by adamkucharski — COMMENTED** [review]
Have added a few small comments on readability

File: posts/share-cpp/index.qmd (hunk 1 above)
Comment (pratikunterwegs) [reply posts/share-cpp/index.qmd:25]: Yes, done in https://github.com/epiverse-trace/epiverse-trace.github.io/pull/34/commits/5fc90a324aa144aba5fd0828d24bb1338f0c8e6f

File: posts/share-cpp/index.qmd (hunk 2 above)
Comment (pratikunterwegs) [reply posts/share-cpp/index.qmd:99]: Hopefully clearer after https://github.com/epiverse-trace/epiverse-trace.github.io/pull/34/commits/8efbb4ed0c36569855d9154a757e99a98af3d77c?

File: posts/share-cpp/index.qmd (hunk 3 above)
Comment (pratikunterwegs) [reply posts/share-cpp/index.qmd:193]: Fixed in https://github.com/epiverse-trace/epiverse-trace.github.io/pull/34/commits/19f9fea49dfa48aa52d007cde29a79615be7d630

File: posts/share-cpp/index.qmd (hunk 4 above)
Comment (pratikunterwegs) [reply posts/share-cpp/index.qmd]: Fixed in https://github.com/epiverse-trace/epiverse-trace.github.io/pull/34/commits/19f9fea49dfa48aa52d007cde29a79615be7d630

**Comment by pratikunterwegs** [comment]
Thanks for the suggestions @adamkucharski, should all be fixed but happy to make more changes as well.

**Review by Bisaloo — CHANGES_REQUESTED** [review]
Thanks, this post is useful and clearly presents the different options. :+1: 

My main comment would be to carefully go through it and remove any instance of [dismissive or demotivating language](https://carpentries.github.io/instructor-training/04-expertise.html#just-and-other-dismissive-language), such as "naturally", "obviously", "easy", etc. I have flagged some of them but others remain.
Similarly, we want to make sure that we explain where the post fits in the current documentation landscape without belittling existing resources.

I'll update the contributing guide in this sense.

**Comment by pratikunterwegs** [comment]
Thanks for the suggestions @Bisaloo - I seem to have missed the notifications completely! - perhaps I was on leave. Will get this done today.

File: posts/share-cpp/index.qmd [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+Defining this header is the key step in making (R)C++ code shareable.
+
+```sh
+# conceptual organisation of an Rcpp package with a package header
+.
+├── DESCRIPTION
+├── NAMESPACE
+├── R
+│   ├── RcppExports.R
+│   └── mypackage-package.R
```
[the remark this answers is not in the document]
Comment (pratikunterwegs) [reply posts/share-cpp/index.qmd]: Removed in https://github.com/epiverse-trace/epiverse-trace.github.io/pull/34/commits/3ce93709f03c732fc3c74d710c6d73a6947ab00a

**Comment by pratikunterwegs** [comment]
Hi @Bisaloo, I've made changes as suggested in https://github.com/epiverse-trace/epiverse-trace.github.io/pull/34/commits/3ce93709f03c732fc3c74d710c6d73a6947ab00a as well as accepting your edits; please let me know if there's anything more I could do.

File: posts/share-cpp/index.qmd [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+
+Any functions defined in the package headers must be inline functions (see the example above).
+This prevents compilation errors related to multiple definitions.
+
+C++ source files should include the package header, using `#include mypackage.h`.
+Functions, structs, or classes defined in header files will be available from the namespace `mypackage`, as shown in the example below.
+
+The code in header files will usually need to be wrapped in (R)C++ code that is exposed to R to make functions from the headers available in R --- see the snippet below.
+
+```c++
```
Comment (Bisaloo) [inline posts/share-cpp/index.qmd:171]: Could you add a filename hint here as well please?

Comment (pratikunterwegs) [reply posts/share-cpp/index.qmd]: Added in https://github.com/epiverse-trace/epiverse-trace.github.io/pull/34/commits/8dd883c7a4a763d2f58e2446bd764bced6488033

**Comment by Bisaloo** [comment]
Thanks!

**Pull request merged**